### DISCUSSION TITLE
Running multiple replicas of Pi-Hole

### DISCUSSION THREAD

**Discussion by larivierec** [MoJo2600/pihole-kubernetes#182] [Q&A] [answered]
Hello everyone,

I was just wondering if anyone was able to have two replicas of pi-hole of available in a single cluster.
If not, no problem. If so, it would be nice to share some insight on how you got it working.

Example setup from my end
```
/c/Users/Chris-PC $ kc get all -n pihole
NAME                          READY   STATUS    RESTARTS   AGE
pod/pihole-7f55ff8467-xblql   2/2     Running   0          13h
pod/pihole-7f55ff8467-qxxwl   2/2     Running   0          13h

NAME                     TYPE           CLUSTER-IP      EXTERNAL-IP   PORT(S)                    AGE
service/pihole-dns-udp   LoadBalancer   10.152.183.76   192.168.2.2   53:31070/UDP               13h
service/pihole-web       ClusterIP      10.152.183.64   <none>        80/TCP,443/TCP,49312/TCP   13h
service/pihole-dns-tcp   LoadBalancer   10.152.183.63   192.168.2.2   53:31268/TCP               13h

NAME                     READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/pihole   2/2     2            2           13h

NAME                                DESIRED   CURRENT   READY   AGE
replicaset.apps/pihole-7f55ff8467   2         2         2       13h
```

Issue ->

![image](https://user-images.githubusercontent.com/3633214/131170938-9cf545ac-7251-4534-a36c-ac8d70f026e9.png)

**Comment by MoJo2600** [comment, marked as answer]
I think the issue is that pihole is not designed to be run in a HA setup. So the frontend backend communication is not working when the backend switches between pod  instances.

**Reply by larivierec** [reply]
Ahhh, that too is a possibility..
Alright ill scale down! Thanks Christian

**Reply by MoJo2600** [reply]
I'm not sure if there is something like a "sticky sessions" feature that could be used. So one client is always accessing the same backend instance?

**Reply by larivierec** [reply]
It's a possibility, I think the proper way would be have a persistant gravity db that all pihole instances can write in. I don't know if that's the proper way to do this though.

**Reply by MoJo2600** [reply]
Yes, but this is not supported by pihole as far as I know

**Reply by abyrne133** [reply]
Hey, first of all thank you for this repo, it's great! I'm using it at home. 

This is an old discussion so I'm not sure it the above still applies or if functionality for multiple web interface instances is now supported. If so, please shout! I've tried replicaCount = 3 and the web interface starts misbehaving (log in accepts password but doesn't progress to next screen, probably session related).
I came across this blog post: https://discourse.pi-hole.net/t/pi-hole-high-availability-with-kubernetes/67505, where two separate deployments are created:

-  deployment one: 1 pod replica, read/write abilities, web interface exposed on port 80, DNS on 53 , 
-  deployment two: x pod replicas, read only abilities, port 53 only exposed. I think this means you'll get the DNS high availability, just not the web interface HA. Fair enough.

Am I correct in saying HA is not supported (for web interface, and perhaps pi-hole) through this helm chart yet? Or has that changed? I can see a replicaCount parameter in the Values file.

**Reply by larivierec** [reply]
yes, Pi-Hole, last I checked is stateful and the SQLite database is partially the culprit in this scenario. 
The UI can't write to the sqlite db with multiple replicas. Even if the sqlite db was constructed per replica, you'd still need them to synchronize.

If you want something stateless (where you can have multiple pods), you'll need to use something like blocky but there's no frontend.

Hope that answers your question.
